### PR TITLE
Allow user to pass auth token via query string

### PR DESCRIPTION
**Pull request by Kami**
This pull request allows user to also pass authentication token via `?x-auth-token` query parameter (in addition to passing the token via header).

This popped up while I worked on the webhooks documentation (#1022). Previously webhooks were unauthenticated, but now they require same authentication as other API methods. This means webhooks won't work with 3rd party services such as Github which only allow you to manipulate URL and query parameters, but not headers.

In the past @enykeev has also asked for this feature. IIRC, Server-Sent Events don't allow you to send custom headers which means we can't authenticate those requests.

### CODE REVIEW

File: st2common/st2common/hooks.py [hunk 1]
```diff
@@ -94,18 +96,34 @@ def _abort_other_errors():
         }), status=500)
 
     @staticmethod
-    def _validate_token(headers):
-        """Validate token"""
-        if 'X_Auth_Token' not in headers:
-            LOG.audit('Token is not found in header.')
+    def _validate_token(request):
```
Comment (Kami) [inline st2common/st2common/hooks.py]: Yeah, duplicated code in auth middleware and hook also makes me unhappy. We should refactor it.

**Comment by manasdk** [comment]
We will also need to generate API keys rather than the ephemeral tokens. I guess we could bump up the TTL on the tokens for now and call them API keys.

**Comment by Kami** [comment]
Yeah, that's a quick hack I used for now for sensor token (long TTL).

**Comment by Kami** [comment]
Pushed a change which refactores duplicated code - #1027

File: CHANGELOG.rst [hunk 2]
```diff
@@ -18,8 +18,10 @@ Docs: http://docks.stackstorm.com/latest
   used by the action runner service. Previous, system's default Python binary was used.
 * Fix a race-condition / bug which would occur when multiple packs are installed at the same time.
   (bug-fix)
-* vars can be defined in the ActionChain. (new-feature)
+* Vars can be defined in the ActionChain. (new-feature)
```
Comment (manasdk) [inline CHANGELOG.rst]: `vars` is a keyword :P

File: st2common/st2common/hooks.py [hunk 3]
```diff
@@ -14,9 +14,11 @@
 # limitations under the License.
 
 import datetime
+import urlparse
```
Comment (manasdk) [inline st2common/st2common/hooks.py]: six.moves to keep py3 compat

File: st2common/st2common/hooks.py [hunk 4]
```diff
@@ -13,16 +13,16 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import datetime
+import urlparse
```
Comment (manasdk) [inline st2common/st2common/hooks.py]: six.moves

**Comment by manasdk** [comment]
+1

File: st2common/st2common/hooks.py (hunk 4 above)
Comment (Kami) [inline st2common/st2common/hooks.py]: I thought so as well, but then I grepped code for `urlparse` and I didn't see it using six.moves, so I thought maybe I remember it wrong that this has been moved in Python 3, lol.

Anyway, yeah, you are right, it has indeed been moved. Will also update other code.

Comment (Kami) [inline st2common/st2common/hooks.py]: On a related note - we should also start running tests with Python 3 (if possible, dunno about dependencies).

Comment (manasdk) [inline st2common/st2common/hooks.py]: Yeah we had tried and tests kept breaking. Eventually we claim out code to be py3 compliant but can't recommend it to be run with py3.

Comment (Kami) [inline st2common/st2common/hooks.py]: urlparse six.moves change in c7d4574012d5cfb519e77992b0e0101a2cf10486

**Pull request merged**